### ISSUE TITLE
Possible memory leak when saving images

### ISSUE THREAD

**Issue by jtorresfabra** [libvips/libvips#1234]
Hi I'm experiencing a memory leak in a process that saves thousands of images: 
This simple code reproduces the issue, you can check memory consumption is never going down. 

The culprit seems to be vips_image_write_to_file, if commented memory seems stable.

```
#include <vips/vips8>

int main(int argc, char* argv[])
{
    if (VIPS_INIT(argv[0]))
        vips_error_exit(NULL);
    for (int var = 0; var < 1000000; ++var) {
        VipsImage* tmp;
        VipsImage* image = vips_image_new_from_file("/path/to/whatever/image", "access", VIPS_ACCESS_RANDOM, nullptr);
        vips_extract_area(image, &tmp, 0, 0, 100, 100,NULL);
        vips_image_write_to_file(tmp, "/tmp/im.jpg", "Q", 88, "profile", "none", NULL);
        g_object_unref(image);
        g_object_unref(tmp);
    }
    vips_shutdown();
    return 0;
}
```

**Comment by jtorresfabra** [comment]
More data: the issue happens when saving png/jpg. But not when saving ".v".

**Comment by lovell** [comment]
Hello, are you able to provide a report from valgrind or similar. Does preventing the use of caching via `vips_cache_set_max(0)` have any effect? Which version of libvips is this with?

> "memory consumption is never going down"

 Please remember that RSS can include memory that has been freed but has yet to be returned to the OS.

**Comment by jtorresfabra** [comment]
Hi, thanks for your answer: 

> Hello, are you able to provide a report from valgrind or similar
On it.

> Which version of libvips is this with?
Current master

> Does preventing the use of caching via vips_cache_set_max(0) have any effect?
Yep, memory consumption is higher, code is very slow compared with cached version, but memory seems stable or at least controled.

> Please remember that RSS can include memory that has been freed but has yet to be returned to the OS.
Well, it happens in my desktop til memory exhaustion (64Gb RAM) and then it crashes. And also what would be the explanation about saving ".v" does not show the issue?

Thanks

**Comment by jcupitt** [comment]
I wouldn't use git master in production -- it's the development version and won't have been tested so much. 

git master has a new feature: support for XMP in PNG. Perhaps that's causing this? I'll experiment.

**Comment by jtorresfabra** [comment]
Hi @jcupitt ,

Not using in production, but as I have the problem in 8.7 I tried with master. So probably not realted to the new feature (also it happens with jpg). Valgrind is not showing nothing meaningful, but smart pointers/ref counting, can be hidding the issue. 

I'm trying myself to understand what's going on. I'll keep you posted.

**Comment by jtorresfabra** [comment]
BTW do you reproduce?

**Comment by jcupitt** [comment]
I tried a few things (various vips versions, memory settings, images) and I always see a pretty steady:

iterations | MB RES
-------- | --------
10000 | 53
20000 | 70
30000 | 85
40000 | 103
50000 | 117
60000 |132

So about 15mb every 10,000 image writes. Is that the kind of memory creep you are seeing?

valgrind shows nothing and the vips refcount checker shows nothing, so I'm inclined to just blame fragmentation.

**Comment by jtorresfabra** [comment]
Ummm as long as I understand there should be no fragmentation as the code just read an image, write to disk and free all before going again. Probably I'm missing something. 

If using a single core I see more or less what you see, but the more number of cores the more leakage I see (here I have 32). I'm  trying to pinpoint the issue but is really hard. 

Also maybe related or maybe not. In debug multithreaded is slower for this code than setting vips_concurrency_set(1). I see that each time all the threads are killed created again.

**Comment by jtorresfabra** [comment]
Hi @jcupitt, 
ok I think I'm starting to understand how this works. 

Correct me if I'm wrong vips_image_write_to_file, or vips_jpegsave returns the control to main thread before the file is totally written, the pool is still working when the g_object_unref is called. I'm debugging and is what actually I am seeing. 

Do we have a way to actually wait til the write is done?

**Comment by jcupitt** [comment]
I'm now wondering if this is a glib issue. I'm seeing a malloc in `g_type_class_ref()` come up more than it should. I'm trying with a debug glib build now.

Re. jpegsave returning before write is done: I think it's supposed to wait, but perhaps there's a problem. I'll have a look.

**Comment by jcupitt** [comment]
My current test program, fwiw:

```C
/* compile with 
 *
 * gcc -g -Wall soak7.c `pkg-config vips --cflags --libs`
 */

#include <vips/vips.h>
#include <vips/vector.h>

int 
main(int argc, char* argv[])
{
    int i;   

    if (VIPS_INIT(argv[0]))
        vips_error_exit(NULL);

    vips_cache_set_max(0);
    vips_vector_set_enabled(FALSE);

    for (i = 0; i < 1000; ++i) {
        VipsImage* tmp;
        VipsImage* image;
        double avg;

        printf("loop %d ..\n", i);
        image = vips_image_new_from_file(argv[1], 
                "access", VIPS_ACCESS_RANDOM, NULL);
        vips_extract_area(image, &tmp, 0, 0, 100, 100, NULL);
        vips_avg(tmp, &avg, NULL);
        g_object_unref(tmp);
        g_object_unref(image);
    }

    vips_shutdown();

    return 0;
}
```

**Comment by jcupitt** [comment]
Here's a clue: with the program above on an 8-thread desktop I see:


iterations | kb RES
-------- | --------
1000 | 173796
2000 | 180060
3000 | 180616
4000 | 182344
5000 | 182920
6000 | 184646
7000 | 185764

If I add `vips_concurrency_set(1);` I see:

iterations | kb RES
-------- | --------
1000 | 35324
2000 | 35324
3000 | 35324
4000 | 35324
5000 | 35324
6000 | 35324
30000 | 35324

Nothing shows up under valgrind, even with `--leak-check=full --show-leak-kinds=all`, so perhaps it's an issue in pthreads?

**Comment by jtorresfabra** [comment]
just to make it clear as I don't see your code saving the image: my problem
is when writing the image to jpg/png. When saving to .v I don't see any
issue.

Also if the problem is with a smart pointer still referenced probably
valgrind won't show anything meaningful.

Also it seemed to me that using set_max_cache(0) memory remained stable.

Also as a last clue I tried to force delete the image after write and
unref. And it showed that writing was still going on, as unreffing didn't
call the dispose image as it should. (Added a breakpoint in that code and
it was not being called).

I am out for the weekend but will keep digging on monday. Thanks so far for
your help.

**Comment by jcupitt** [comment]
I checked the write after save, and it seems OK here. I added a printf() to vips2jpeg.c here (where the write() to the jpg file happens):

https://github.com/libvips/libvips/blob/master/libvips/foreign/vips2jpeg.c#L445

and here, where the loop returns:

https://github.com/libvips/libvips/blob/master/libvips/foreign/vips2jpeg.c#L649

I tried 1000 iterations and I always see:

```
$ ./a.out ~/pics/k2.jpg | grep -A 1 vips_sink_disc x | grep loop | wc
    999    2997   11880
```

ie. every sink_disc is followed by another `loop` and never by a delayed `write_jpeg_block()`.

unref won't always trigger dispose, it depends on the cache.

**Comment by jcupitt** [comment]
Yes, I removed the image save as I found I could still see slight memory creep even without it. `avg` is a nice simple sink, so I hoped it would make valgrind output easier to read.

**Comment by jtorresfabra** [comment]
Ok after sweating a bit I have some advances. 

Not fully fixed but it seems the thread pool is creating each thread and each ThreadPool associated to the pool->im.
https://github.com/libvips/libvips/blob/master/libvips/iofuncs/threadpool.c#L706
https://github.com/libvips/libvips/blob/master/libvips/iofuncs/threadpool.c#L827

As the cache does not let the image to be freed, it never calls the postclose callback, which should do the call to g_free. (https://github.com/libvips/libvips/blob/master/libvips/iofuncs/memory.c#L180)

So, if instead of creating the threads and the threadpool associated to the image I do it with (NULL, ...)  and call the respective g_free in the thread destruction places, I get a nice reduction of the leak. Of course this is not viewable by valgrind cause g_object is hiding the issue. So as for each image I want to save a new threadpool is created, so the more threads the bigger the leak is.  

Hope it makes sense.

**Comment by jtorresfabra** [comment]
Hi @jcupitt 

ok so I got a 0 leakage finally, would you accept a PR on this?

**Comment by jcupitt** [comment]
Oh, interesting! 

But what do you mean when you say "As the cache does not let the image to be freed"? The cache will free the image eventually, and you can turn the case off, of course. Do you think there is a circular reference, perhaps? I'll check.

> I was surprised that despite setting vips_concurrency to 1 you can have more than one thread running, cause https://github.com/libvips/libvips/blob/master/libvips/iofuncs/sinkdisc.c#L235 does not take into account the concurrency limits.

That's right, the write-behind threads are not part of the threadpool. There are always exactly two extra write-behind threads no matter what size the worker pool is.

**Comment by jtorresfabra** [comment]
>But what do you mean when you say "As the cache does not let the image to be freed"? The cache will free the image eventually, and you can turn the case off, of course. Do you think there is a circular reference, perhaps? 

Yeah it will free the image eventually, but while it is not freed the code keeps adding threads and pools locally to the object for every sink operation. So yes, once the image is freed then it will free al the objects in the postclose call. But the cache will not likely free the image if you keep doing things with it til you finish, and maybe your memory is already exhausted and the program crashed at that point...

That's why valgrind does not see any leak, because at the end all the memory is freed, but is the typical hidden leak from reference counting. 

Does it make sense now?

**Comment by jcupitt** [comment]
Ah I think I understand. The case you fix is like this:

```C
VipsImage *image = vips_image_new_from_file ("somefile.jpg")l
int i;

for (i = 0; i < 10000; i++) {
  double x;
  
  vips_avg(image, &x, NULL);
}

g_object_unref (image);
```

Now you are running 10000 threadpools over `image`, but they are allocated on `image`, so they won't get freed until `image` is unreffed, which is not until all 10000 have run. So memory will climb during the 10,000 loop.

**Comment by jcupitt** [comment]
Oh heh we crossed over. 

It's not a leak, in my opinion, but you're right that memory is not being freed as early as it could be. Yes, let's do your patch!

**Comment by jcupitt** [comment]
Also, well done for tracking this down! Let's move to your PR.

**Comment by jtorresfabra** [comment]
Maybe a terminology problem, I used to name that a hidden leak, cause it can clearly cause a crash for memory exhaustion when you run programs for a long time or in resource limited devices.

I already faced this several times with smart pointers and they are quite hard to find.

Thanks for your patience!
